### ISSUE TITLE
dart compile exe support static linking of C code

### ISSUE THREAD

**Issue by dcharkes** [dart-lang/sdk#47718]
`dart compile exe` enables compiling Dart source code to a standalone executable.

It would be nice if we could link in static libraries in this process so that we can use `DynamicLibrary.executable()` to look up the symbols.

Current workaround: Use shared libraries, use `DynamicLibrary.open()` with a relative path, and ship a folder with the `dart compile exe` executable and the shared libraries (`*.so` / `*.dll` / `*.dylib`).

Technical work that needs to happen for this:

- [ ] Make `dart compile exe` use a proper format instead of appending the precompiled programing as a blob to the dart precompiled runtime by using a a proper linker on each target platform.
- [ ] Add options to pass in linker arguments.

(cc @mkustermann, @cskau-g I don't think we had an issue tracking this yet.)

**Comment by dvc94ch** [comment]
was just trying to figure out how to do this, this would be great, yes.

**Comment by xros** [comment]
Yes. That's what we need. Compile C/C++ libs together into a standalone executable.

I come up with an idea that. In the future we can have a function called `StaticLibrary.open()`.
And after that, everytime when we try to compile it, dart compiler can add the C/C++ libs to the end of the main binary. When the binary is executed, it would look for the place of the C/C++ lib where was attached to itself. It would have some affect on loading time, and RAM consumption. But it's easier to deliver apps. Just make it optional.

This feature can be only for `dart compile exe`, not for `aot-snapshot`, `jit-snapshot`, `kernel`.

**Comment by dcharkes** [comment]
> Make `dart compile exe` use a proper format instead of appending the precompiled program as a blob to the dart precompiled runtime by using a a proper linker on each target platform.

Work underway for MacOS: https://dart-review.googlesource.com/c/sdk/+/228080 

cc @sstrickl

**Comment by dvc94ch** [comment]
Also please make the linker configurable, cross linking with lld works on all platforms these days and would be nice to support in addition/instead of the system linker.

**Comment by dcharkes** [comment]
We would like to address this using the static bindings with `@FfiNative`s:

* https://github.com/dart-lang/sdk/issues/49418

**Comment by maks** [comment]
@laoshaw `dart compile` *already* produces a single, executable without any dependencies except the usual C runtime ones:
```
$ dart compile exe main.dart 
$ ldd main.exe 
        linux-vdso.so.1 (0x00007fff4a3fc000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f63d4122000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f63d411d000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f63d4036000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f63d3e0e000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f63d47ea000)
```

This issue is specifically about being able to do so when also link in static libraries into that executable *when* using Dart FFI.

**Comment by dcharkes** [comment]
This issue tracks statically linking in C libraries that package developers provide.

Statically linking in the C runtime dependencies of the dart aot runtime would be a different project. @laoshaw do you have any issues with deployment depending on the C runtime libraries that are available? These libraries are always pre-installed on Linux. And bundling them would significantly increase the binary size of your final program.

**Comment by mkustermann** [comment]
> Yes, I would like ldd main.exe outputs not a dynamic executable instead of listing all those C libraries, that makes depolyments as simple as golang, is this possible?

Doing this may already be a showstopper as it may break a number of things, see e.g. [discussion here](https://stackoverflow.com/questions/57476533/why-is-statically-linking-glibc-discouraged). I also recall that such statically linked binaries will have trouble opening dynamic libraries, which is something `dart:ffi` provides and we cannot break.

**Comment by rusty-snake** [comment]
https://dart.dev/tools/dart-compile:
> The exe subcommand produces a **standalone executable** for Windows, macOS, or Linux. A standalone executable is native machine code that's compiled from the specified Dart file and its dependencies, plus a small [Dart runtime](https://dart.dev/overview#runtime) that handles type checking and garbage collection.
>
> […]
>
> Known limitations
> The exe subcommand has some known limitations:
>
> - No cross-compilation support
> - No support for dart:mirrors and dart:developer

A *standalone* executable that depended on a glibc with correct version being present at runtime is not what I expect when I hear "*standalone* executable".

Given that this will not be "fixed" any time soon (no activity for 1.5 years) this should be clarified / added to known limitations.

**Comment by chances** [comment]
@rusty-snake Much to my chagrin, there are tons of "standalone" applications in the wild that depend on a specific C library being installed to function. Regardless, this issue is specific to user-supplied C libraries at compile-time.

@dcharkes Consider renaming this issue to reduce confusion.

**Comment by RogueZamboni** [comment]
Is there any hope of this ever being implemented?   We would like to adopt dart on a couple of projects, but not having this is a deal-breaker.

**Comment by dcharkes** [comment]
> Is there any hope of this ever being implemented? We would like to adopt dart on a couple of projects, but not having this is a deal-breaker.

@RogueZamboni Could you elaborate on your use case, why doesn't dynamic loading of C code work? With `dart compile cli` you'll get an app bundle that contains a `.exe` and the dynamic libraries. Please see the documentation for dynamic loading of C code here: https://dart.dev/tools/hooks

**Comment by ryanheise** [comment]
Since this hasn't been answered yet, I'll suggest an answer: App bundles are fine for some platforms, particularly app stores where this is the norm and built into the platform. Other platforms like Linux have no standard app bundle installation process that every user is happy with, so a single file downloadable executable can be an attractive option for someone who is looking for simplicity. There are probably other minor reasons, such as the simplification of deploy-to-server scripts - one file is easier than multiple files. I recognise that shared libraries can work in both situations, but a single file executable is still a nice option to have.

**Comment by maks** [comment]
I'll second @ryanheise answer as especially for backend deployments vs GUI apps, a single executable has *alot* of benefits and this is used to very good effect in the Go ecosystem, eg. [PocketBase](https://github.com/pocketbase/pocketbase/releases).  

Likewise for other use cases like a static site generator, which I for instance maintain and find very useful to have [as a single binary](https://github.com/maks/picosite/releases).

**Comment by RogueZamboni** [comment]
I agree with both of the answers given here, but allow me to add an arguably somewhat silly reason:  for certain security purposes or in some infrastructure scenarios each loadable binary artifact must go through a somewhat intense and costly process and some stakeholders will multiply their cost in doing so by the number of artifacts, not the size / number of symbols / etc.   You can see a somewhat similar thing happening in enterprises that use per-binary whitelists in their security infrastructure, if not a stringently accounted one.

In general:  people who are already producing deliverables as single-binary for whatever reason they are, with users who are used to that and depend on it will not likely easily migrate to Dart if it means they need to stop doing that.  Static compilation is almost universally supported in general-purpose language compilers / code generators.   I get that Dart grew up fit to purpose for an ecosystem in which "app bundles" just make sense, but that's a niche relative to "general purpose language".

In any case I'm not really looking to be sold on app bundles or discuss much about &lt;X&gt;/&lt;Y&gt;, I just want to know if support for static is conceivably on the horizon somewhere, so I can better allocate my efforts.
